query.go

```diff
@@ -851,31 +712,89 @@ func (s *ChainService) GetCFilter(blockHash chainhash.Hash,
 	// With all the necessary items retrieved, we'll launch our concurrent
 	// query to the set of connected peers.
 	log.Debugf("Fetching filters for heights=[%v, %v], stophash=%v",
-		query.startHeight, query.stopHeight, query.stopHash)
+		q.startHeight, q.stopHeight, q.stopHash)
 
 	go func() {
 		defer s.mtxCFilter.Unlock()
-		defer close(query.filterChan)
-
-		s.queryPeers(
-			// Send a wire.MsgGetCFilters.
-			query.queryMsg(),
-
-			// Check responses and if we get one that matches, end
-			// the query early.
-			func(_ *ServerPeer, resp wire.Message, quit chan<- struct{}) {
-				s.handleCFiltersResponse(query, resp, quit)
-			},
-			query.options...,
+
+		// Hand the query to the work manager, and consume the verified
+		// responses as they come back.
+		errChan := s.queryDispatcher.Query(
+			[]*query.Request{q.request()}, query.Cancel(s.quit),
+			query.Encoding(qo.encoding),
+		)
+
+		var (
+			resp *filterResponse
+			ok   bool
 		)
 
-		// If there are elements left to receive, the query failed.
-		if len(query.headerIndex) > 0 {
-			numFilters := query.stopHeight - query.startHeight + 1
-			numRecv := numFilters - int64(len(query.headerIndex))
-			log.Errorf("Query failed with %d out of %d filters "+
-				"received", numRecv, numFilters)
-			return
+		for {
+			select {
+			case resp, ok = <-q.filterChan:
```

Comment:
Hmm, I think better to not have the select assignment override the variable above? Also given that `ok` isn't used in the scope below. So can use an intermediate variable.